File: packages/sdk-python/transparentguard/pii.py
"""
TransparentGuard Python SDK — PII Detector
Regex-based personally identifiable information detection.
Covers the same PII categories as the TypeScript runtime.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Set

from .types import PiiCategory, PiiTarget


# ---------------------------------------------------------------------------
# PII Match
# ---------------------------------------------------------------------------

@dataclass
class PiiMatch:
    category: str
    start: int
    end: int
    original: str
    confidence: float = 1.0


# ---------------------------------------------------------------------------
# Regex patterns per category
# ---------------------------------------------------------------------------

_PATTERNS: Dict[str, List[re.Pattern[str]]] = {
    "email": [
        re.compile(
            r"\b[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}\b",
            re.IGNORECASE,
        )
    ],
    "phone": [
        re.compile(r"\b(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b"),
        re.compile(r"\b\+[1-9]\d{7,14}\b"),
    ],
    "ssn": [
        re.compile(r"\b(?!000|666|9\d{2})\d{3}[-\s]?(?!00)\d{2}[-\s]?(?!0000)\d{4}\b"),
    ],
    "credit_card": [
        re.compile(r"\b(?:4\d{12}(?:\d{3})?|5[1-5]\d{14}|3[47]\d{13}|3(?:0[0-5]|[68]\d)\d{11}|6(?:011|5\d{2})\d{12}|(?:2131|1800|35\d{3})\d{11})\b"),
        re.compile(r"\b\d{4}[-\s]\d{4}[-\s]\d{4}[-\s]\d{4}\b"),
    ],
    "ip_address": [
        re.compile(r"\b(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\b"),
        re.compile(
            r"\b(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}\b|"
            r"\b(?:[0-9a-fA-F]{1,4}:)*::(?:[0-9a-fA-F]{1,4}:)*[0-9a-fA-F]{1,4}\b"
        ),
    ],
    "iban": [
        re.compile(r"\b[A-Z]{2}\d{2}[A-Z0-9]{4}\d{7}(?:[A-Z0-9]?){0,16}\b"),
    ],
    "passport": [
        re.compile(r"\b[A-Z]{1,2}\d{6,9}\b"),
        re.compile(r"\bpassport\s*(?:number|no\.?|#)?\s*:?\s*([A-Z]{1,2}\d{6,9})\b", re.IGNORECASE),
    ],
    "driver_license": [
        re.compile(r"\b(?:driver(?:'?s)?\s*licen[cs]e|dl|d\.l\.)\s*(?:number|no\.?|#)?\s*:?\s*([A-Z0-9]{6,15})\b", re.IGNORECASE),
    ],
    "bank_account": [
        re.compile(r"\b(?:account|acct)\.?\s*(?:number|no\.?|#)?\s*:?\s*\d{8,17}\b", re.IGNORECASE),
    ],
    "mrn": [
        re.compile(r"\bMRN[-:\s]?\d{6,10}\b", re.IGNORECASE),
        re.compile(r"\bmedical\s+record\s+(?:number|no\.?|#)?\s*:?\s*\d{6,10}\b", re.IGNORECASE),
    ],
    "dob": [
        re.compile(r"\b(?:DOB|date\s+of\s+birth)\s*:?\s*\d{1,2}[\/\-\.]\d{1,2}[\/\-\.]\d{2,4}\b", re.IGNORECASE),
        re.compile(r"\b(?:born|birth(?:day)?)\s+(?:on\s+)?\d{1,2}[\/\-\.]\d{1,2}[\/\-\.]\d{2,4}\b", re.IGNORECASE),
    ],
    "name": [
        # Contextual name detection — looks for "Name: John Smith" patterns
        re.compile(r"\b(?:patient|client|customer|user|name)\s*:?\s*([A-Z][a-z]+(?:\s+[A-Z][a-z]+){1,3})\b"),
        # Full name in PII context (e.g., "I am John Smith" in medical context)
        re.compile(r"\b(?:I am|my name is|called)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+){1,2})\b"),
    ],
    "address": [
        re.compile(
            r"\b\d{1,5}\s+[A-Za-z0-9\s,\.]{3,50}"
            r"(?:Street|St|Avenue|Ave|Boulevard|Blvd|Road|Rd|Lane|Ln|Drive|Dr|Way|Court|Ct|Place|Pl)"
            r"(?:\s*,\s*(?:Suite|Ste|Apt|Unit|#)\s*[\w\d]+)?"
            r"(?:\s*,\s*[A-Za-z\s]+)?"
            r"(?:\s*,\s*[A-Z]{2})?"
            r"(?:\s+\d{5}(?:-\d{4})?)?\b",
            re.IGNORECASE,
        )
    ],
    "crypto_address": [
        # Bitcoin
        re.compile(r"\b[13][a-km-zA-HJ-NP-Z1-9]{25,34}\b"),
        # Ethereum
        re.compile(r"\b0x[a-fA-F0-9]{40}\b"),
    ],
    "tax_id": [
        re.compile(r"\b(?:EIN|tax\s+id)\s*:?\s*\d{2}-\d{7}\b", re.IGNORECASE),
    ],
    "health_condition": [
        re.compile(
            r"\b(?:diagnosed with|suffers? from|has|history of|treatment for)\s+"
            r"(?:diabetes|cancer|HIV|AIDS|hepatitis|depression|anxiety|bipolar|schizophrenia"
            r"|hypertension|heart disease|epilepsy|Alzheimer'?s|dementia|ADHD|autism"
            r"|lupus|multiple sclerosis|Parkinson'?s|PTSD|OCD)\b",
            re.IGNORECASE,
        )
    ],
    "national_id": [
        re.compile(r"\b(?:national\s+id|NIN|national\s+insurance)\s*(?:number|no\.?|#)?\s*:?\s*[A-Z0-9]{6,15}\b", re.IGNORECASE),
    ],
    "npi": [
        re.compile(r"\bNPI\s*(?:number|no\.?|#)?\s*:?\s*\d{10}\b", re.IGNORECASE),
    ],
}

# ---------------------------------------------------------------------------
# Category expansions (logical groups)
# ---------------------------------------------------------------------------

_CATEGORY_EXPANSION: Dict[str, List[str]] = {
    "pii_all": [
        "name", "email", "phone", "address", "ip_address", "ssn", "passport",
        "driver_license", "national_id", "tax_id", "credit_card", "bank_account",
        "iban", "crypto_address", "mrn", "dob", "health_condition", "npi",
        "race", "religion", "political_opinion", "sexual_orientation",
        "biometric", "genetic", "union_membership",
    ],
    "phi": [
        "name", "address", "phone", "email", "dob", "ssn", "mrn",
        "health_condition", "insurance_id", "npi", "dea",
    ],
    "pii_standard": [
        "name", "email", "phone", "address", "ip_address", "username",
    ],
    "pii_financial": [
        "credit_card", "bank_account", "iban", "swift", "crypto_address", "tax_id",
    ],
    "pii_sensitive": [
        "ssn", "passport", "driver_license", "national_id", "voter_id",
        "mrn", "health_condition", "biometric", "genetic",
        "race", "religion", "political_opinion", "sexual_orientation", "union_membership",
    ],
}


def expand_categories(categories: List[str]) -> Set[str]:
    """Expands logical category groups to their constituent categories."""
    expanded: Set[str] = set()
    for cat in categories:
        if cat in _CATEGORY_EXPANSION:
            expanded.update(_CATEGORY_EXPANSION[cat])
        else:
            expanded.add(cat)
    return expanded


# ---------------------------------------------------------------------------
# Detection
# ---------------------------------------------------------------------------

def detect_pii(text: str, target: PiiTarget) -> List[PiiMatch]:
    """
    Detects PII in the given text for the specified PiiTarget configuration.
    Returns a list of PiiMatch objects sorted by start position.
    """
    confidence_threshold = target.get("confidence_threshold", 0.70)
    categories = expand_categories(list(target.get("categories", [])))

    matches: List[PiiMatch] = []
    seen: Set[tuple[int, int]] = set()

    for category in categories:
        patterns = _PATTERNS.get(category, [])
        for pattern in patterns:
            for m in pattern.finditer(text):
                span = (m.start(), m.end())
                if span in seen:
                    continue
                # Assign confidence based on match quality, checksums, and context
                confidence = _confidence_for_category(category, m.group(0), text)
                if confidence < confidence_threshold:
                    continue
                seen.add(span)
                matches.append(PiiMatch(
                    category=category,
                    start=m.start(),
                    end=m.end(),
                    original=m.group(0),
                    confidence=confidence,
                ))

    matches.sort(key=lambda x: x.start)
    return _remove_overlapping(matches)


def _luhn_check(number: str) -> bool:
    """Luhn algorithm checksum — validates credit card numbers."""
    digits = [int(c) for c in number if c.isdigit()]
    if len(digits) < 13:
        return False
    total = 0
    for i, d in enumerate(reversed(digits)):
        if i % 2 == 1:
            d *= 2
            if d > 9:
                d -= 9
        total += d
    return total % 10 == 0


def _ssn_valid_range(digits: str) -> bool:
    """Rejects known-invalid SSN area numbers (000, 666, 900-999)."""
    if len(digits) < 3:
        return False
    area = int(digits[:3])
    return area not in (0, 666) and not (900 <= area <= 999)


def _confidence_for_category(category: str, match_text: str, full_text: str = "") -> float:
    """
    Computes a dynamic confidence score for a PII match.

    Signals (additive, clamped to [0.0, 1.0]):
    - Base score per category precision tier
    - +0.08 for labelled match (PII keyword adjacent in context)
    - +0.08 checksum bonus for credit_card/ssn passing validation
    - -0.15 penalty (capped at 0.60) if checksum fails
    - +0.05 if PII-adjacent context words appear within 100 chars of match
    """
    high_precision = {
        "email", "credit_card", "ssn", "iban", "ip_address",
        "crypto_address", "mrn", "npi", "tax_id",
    }
    medium_precision = {
        "phone", "passport", "bank_account", "dob", "driver_license",
        "national_id", "health_condition",
    }

    if category in high_precision:
        base = 0.88
    elif category in medium_precision:
        base = 0.76
    else:
        base = 0.65  # name, address, keyword categories

    score = base

    # Labelled match: PII keyword appears just before the match in full text
    if full_text:
        match_pos = full_text.find(match_text)
        if match_pos > 0:
            pre_window = full_text[max(0, match_pos - 80):match_pos].lower()
            label_keywords = [
                "ssn", "passport", "dob", "routing", "ein", "tin", "iban",
                "mrn", "npi", "credit card", "credit_card", "driver", "license",
                "account", "member id", "patient", "date of birth",
            ]
            if any(kw in pre_window for kw in label_keywords):
                score += 0.08

            # Context proximity: PII-adjacent nouns near the match
            window_start = max(0, match_pos - 100)
            window_end   = min(len(full_text), match_pos + len(match_text) + 100)
            context = full_text[window_start:window_end].lower()
            context_words = [
                "patient", "account", "billing", "insurance", "member",
                "medical record", "diagnosis", "prescription", "routing",
                "social security", "birth", "license", "passport",
            ]
            if any(w in context for w in context_words):
                score += 0.05

    # Checksum / range validation
    clean = re.sub(r"[\s\-]", "", match_text)
    if category == "credit_card":
        if _luhn_check(clean):
            score += 0.08
        else:
            score -= 0.15
            score = min(score, 0.60)
    elif category == "ssn":
        digits_only = re.sub(r"\D", "", clean)
        if _ssn_valid_range(digits_only):
            score += 0.08
        else:
            score -= 0.15
            score = min(score, 0.60)

    return max(0.0, min(1.0, score))


def _remove_overlapping(matches: List[PiiMatch]) -> List[PiiMatch]:
    """Removes overlapping matches, keeping the one with higher confidence."""
    if not matches:
        return []
    result = [matches[0]]
    for match in matches[1:]:
        last = result[-1]
        if match.start < last.end:
            # Overlap — keep the one with higher confidence
            if match.confidence > last.confidence:
                result[-1] = match
        else:
            result.append(match)
    return result


# ---------------------------------------------------------------------------
# Redaction
# ---------------------------------------------------------------------------

def redact_text(text: str, matches: List[PiiMatch]) -> str:
    """
    Applies redaction to the given text, replacing each matched span
    with [REDACTED:<CATEGORY>].
    Processes matches in reverse order to preserve character positions.
    """
    result = text
    for match in reversed(sorted(matches, key=lambda m: m.start)):
        replacement = f"[REDACTED:{match.category.upper()}]"
        result = result[:match.start] + replacement + result[match.end:]
    return result

File: packages/runtime/src/evaluators/pii.ts
/**
 * TransparentGuard Runtime — PII Evaluator (Free Tier)
 * Pure regex-based detection covering all TPS v1.0 PII categories.
 * No external dependencies — runs entirely offline.
 */

import type { PiiCategory, PiiTarget } from "../types.js";

// ---------------------------------------------------------------------------
// PII category expansion — shortcut aliases
// ---------------------------------------------------------------------------

const PHI_CATEGORIES: PiiCategory[] = [
  "name", "email", "phone", "address", "ssn", "mrn", "dob", "age",
  "health_condition", "insurance_id", "npi", "dea", "ip_address",
  "device_id", "url", "biometric", "genetic",
];

const PII_STANDARD_CATEGORIES: PiiCategory[] = [
  "name", "email", "phone", "address", "ip_address", "username", "url",
];

const PII_FINANCIAL_CATEGORIES: PiiCategory[] = [
  "credit_card", "bank_account", "iban", "swift", "crypto_address", "tax_id",
];

const PII_SENSITIVE_CATEGORIES: PiiCategory[] = [
  "ssn", "passport", "driver_license", "national_id", "tax_id", "voter_id",
  "race", "religion", "political_opinion", "sexual_orientation",
  "biometric", "genetic", "union_membership",
];

const PII_ALL_CATEGORIES: PiiCategory[] = [
  "name", "email", "phone", "address", "ip_address", "username", "device_id", "url",
  "ssn", "passport", "driver_license", "national_id", "tax_id", "voter_id",
  "credit_card", "bank_account", "iban", "swift", "crypto_address",
  "mrn", "dob", "age", "health_condition", "insurance_id", "npi", "dea",
  "race", "religion", "political_opinion", "sexual_orientation",
  "biometric", "genetic", "union_membership",
];

export function expandCategories(categories: PiiCategory[]): PiiCategory[] {
  const expanded = new Set<PiiCategory>();
  for (const cat of categories) {
    switch (cat) {
      case "phi":
        PHI_CATEGORIES.forEach((c) => expanded.add(c));
        break;
      case "pii_standard":
        PII_STANDARD_CATEGORIES.forEach((c) => expanded.add(c));
        break;
      case "pii_financial":
        PII_FINANCIAL_CATEGORIES.forEach((c) => expanded.add(c));
        break;
      case "pii_sensitive":
        PII_SENSITIVE_CATEGORIES.forEach((c) => expanded.add(c));
        break;
      case "pii_all":
        PII_ALL_CATEGORIES.forEach((c) => expanded.add(c));
        break;
      default:
        expanded.add(cat);
    }
  }
  return [...expanded];
}

// ---------------------------------------------------------------------------
// Detection result
// ---------------------------------------------------------------------------

export interface PiiMatch {
  category: PiiCategory;
  start: number;
  end: number;
  original: string;
  confidence: number;
}

// ---------------------------------------------------------------------------
// Regex patterns per category
// Each entry is an array of patterns — first match wins per position.
// All patterns are constructed fresh per call (stateless, no lastIndex issues).
// ---------------------------------------------------------------------------

type PatternEntry = { pattern: string; flags: string; confidence: number };

const PATTERNS: Partial<Record<PiiCategory, PatternEntry[]>> = {
  email: [
    {
      pattern: "[a-zA-Z0-9._%+\\-]+@[a-zA-Z0-9.\\-]+\\.[a-zA-Z]{2,}",
      flags: "gi",
      confidence: 0.98,
    },
  ],

  phone: [
    // +1 (555) 555-5555 or +1-555-555-5555 or (555) 555-5555 etc.
    {
      pattern:
        "(?:\\+?1[\\s\\-.]?)?(?:\\(?\\d{3}\\)?[\\s\\-.]?)\\d{3}[\\s\\-.]\\d{4}",
      flags: "g",
      confidence: 0.90,
    },
    // International formats +44 20 7946 0958
    {
      pattern: "\\+\\d{1,3}[\\s\\-.]\\d{1,4}[\\s\\-.]\\d{3,4}[\\s\\-.]\\d{3,4}",
      flags: "g",
      confidence: 0.85,
    },
  ],

  ssn: [
    // 123-45-6789 (not 000, 666, or 900-999 in first segment — real validation)
    {
      pattern: "(?!000|666|9\\d{2})\\d{3}-(?!00)\\d{2}-(?!0000)\\d{4}",
      flags: "g",
      confidence: 0.97,
    },
    // 123456789 (9 consecutive digits — lower confidence)
    {
      pattern: "\\b(?!000000000)(?!111111111)\\d{9}\\b",
      flags: "g",
      confidence: 0.65,
    },
  ],

  credit_card: [
    // Visa, MC, Amex, Discover — with optional separators
    {
      pattern:
        "\\b(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|6(?:011|5[0-9]{2})[0-9]{12})\\b",
      flags: "g",
      confidence: 0.95,
    },
    // Same with dashes/spaces
    {
      pattern:
        "\\b(?:4[0-9]{3}[\\s\\-]){3}[0-9]{4}|(?:5[1-5][0-9]{2}|2[2-7][0-9]{2})[\\s\\-][0-9]{4}[\\s\\-][0-9]{4}[\\s\\-][0-9]{4}\\b",
      flags: "g",
      confidence: 0.92,
    },
  ],

  bank_account: [
    // Generic US bank account: 6–17 digits
    { pattern: "\\b\\d{6,17}\\b", flags: "g", confidence: 0.60 },
  ],

  iban: [
    {
      pattern: "\\b[A-Z]{2}\\d{2}[A-Z0-9]{4,30}\\b",
      flags: "g",
      confidence: 0.92,
    },
  ],

  swift: [
    // SWIFT/BIC: 8 or 11 characters
    { pattern: "\\b[A-Z]{4}[A-Z]{2}[A-Z0-9]{2}(?:[A-Z0-9]{3})?\\b", flags: "g", confidence: 0.85 },
  ],

  crypto_address: [
    // Bitcoin P2PKH/P2SH
    { pattern: "\\b[13][a-km-zA-HJ-NP-Z1-9]{25,34}\\b", flags: "g", confidence: 0.88 },
    // Ethereum
    { pattern: "\\b0x[a-fA-F0-9]{40}\\b", flags: "g", confidence: 0.96 },
    // Bech32 (bc1...)
    { pattern: "\\bbc1[a-zA-HJ-NP-Z0-9]{25,39}\\b", flags: "g", confidence: 0.93 },
  ],

  ip_address: [
    // IPv4
    {
      pattern:
        "\\b(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\b",
      flags: "g",
      confidence: 0.95,
    },
    // IPv6
    {
      pattern:
        "\\b(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}\\b",
      flags: "gi",
      confidence: 0.95,
    },
  ],

  mrn: [
    // Medical Record Number — typically 6–10 digits, sometimes prefixed
    { pattern: "\\bMRN[:\\s#]*[0-9]{5,10}\\b", flags: "gi", confidence: 0.90 },
    { pattern: "\\bMed(?:ical)?\\s*Record\\s*(?:Number|#|No)[:\\s]*[0-9]{5,10}\\b", flags: "gi", confidence: 0.88 },
  ],

  npi: [
    // National Provider Identifier — 10 digits starting with 1 or 2
    { pattern: "\\bNPI[:\\s#]*[12][0-9]{9}\\b", flags: "gi", confidence: 0.92 },
    { pattern: "\\b[12]\\d{9}\\b", flags: "g", confidence: 0.60 },
  ],

  dea: [
    // DEA Registration Number: 2 letters + 7 digits
    { pattern: "\\b[A-Za-z]{2}[0-9]{7}\\b", flags: "g", confidence: 0.75 },
    { pattern: "\\bDEA[:\\s#]*[A-Za-z]{2}[0-9]{7}\\b", flags: "gi", confidence: 0.95 },
  ],

  passport: [
    // US passport: letter + 8 digits
    { pattern: "\\b[A-Z][0-9]{8}\\b", flags: "g", confidence: 0.72 },
    { pattern: "\\bPassport[:\\s#]*[A-Z0-9]{6,12}\\b", flags: "gi", confidence: 0.88 },
  ],

  driver_license: [
    { pattern: "\\bDL[:\\s#]*[A-Z0-9]{6,12}\\b", flags: "gi", confidence: 0.88 },
    { pattern: "\\bDriver(?:'?s)?\\s*Licen[sc]e[:\\s#]*[A-Z0-9]{6,12}\\b", flags: "gi", confidence: 0.87 },
  ],

  national_id: [
    { pattern: "\\bNational\\s*ID[:\\s#]*[A-Z0-9]{6,20}\\b", flags: "gi", confidence: 0.85 },
  ],

  tax_id: [
    // EIN: XX-XXXXXXX
    { pattern: "\\b\\d{2}-\\d{7}\\b", flags: "g", confidence: 0.88 },
    { pattern: "\\b(?:EIN|TIN|ITIN)[:\\s#]*\\d{2}-?\\d{7}\\b", flags: "gi", confidence: 0.95 },
  ],

  insurance_id: [
    { pattern: "\\bInsurance\\s*(?:ID|Number|#)[:\\s]*[A-Z0-9]{6,20}\\b", flags: "gi", confidence: 0.85 },
    { pattern: "\\bMember\\s*ID[:\\s]*[A-Z0-9]{6,15}\\b", flags: "gi", confidence: 0.78 },
  ],

  dob: [
    // MM/DD/YYYY or MM-DD-YYYY or YYYY-MM-DD
    {
      pattern:
        "\\b(?:0?[1-9]|1[0-2])[/\\-](?:0?[1-9]|[12][0-9]|3[01])[/\\-](?:19|20)\\d{2}\\b",
      flags: "g",
      confidence: 0.88,
    },
    {
      pattern: "\\b(?:19|20)\\d{2}-(?:0[1-9]|1[0-2])-(?:0[1-9]|[12][0-9]|3[01])\\b",
      flags: "g",
      confidence: 0.92,
    },
    // "Date of Birth" label near a date
    {
      pattern:
        "(?:Date\\s*of\\s*Birth|DOB|D\\.O\\.B)[:\\s]*(?:0?[1-9]|1[0-2])[/\\-](?:0?[1-9]|[12][0-9]|3[01])[/\\-](?:19|20)\\d{2}",
      flags: "gi",
      confidence: 0.97,
    },
  ],

  age: [
    { pattern: "\\bage[:\\s]+(?:is\\s+)?([0-9]{1,3})\\s*(?:years?\\s*old)?\\b", flags: "gi", confidence: 0.75 },
    { pattern: "\\b([0-9]{1,3})\\s*(?:yr|year)s?(?:\\s*old)?\\b", flags: "gi", confidence: 0.65 },
  ],

  url: [
    {
      pattern: "https?://[a-zA-Z0-9\\-._~:/?#\\[\\]@!$&'()*+,;=%]+",
      flags: "gi",
      confidence: 0.98,
    },
  ],

  username: [
    { pattern: "\\b@[a-zA-Z0-9_]{2,30}\\b", flags: "g", confidence: 0.80 },
    { pattern: "\\busername[:\\s]+[a-zA-Z0-9_.\\-]{2,30}\\b", flags: "gi", confidence: 0.88 },
  ],

  address: [
    // Street address: number + street name
    {
      pattern:
        "\\b\\d{1,5}\\s+(?:[A-Z][a-z]+\\s+){1,3}(?:Street|St|Avenue|Ave|Road|Rd|Boulevard|Blvd|Drive|Dr|Lane|Ln|Way|Court|Ct|Place|Pl|Circle|Cir)(?:\\s*,\\s*[A-Z][a-z]+(?:\\s+[A-Z][a-z]+)*)?(?:\\s*,\\s*[A-Z]{2})?(?:\\s+\\d{5}(?:-\\d{4})?)?\\b",
      flags: "g",
      confidence: 0.85,
    },
  ],

  name: [
    // "Name: John Doe" patterns — low confidence standalone, higher with label
    { pattern: "(?:Name|Patient|Customer|User)[:\\s]+([A-Z][a-z]+(?:\\s+[A-Z][a-z]+){1,2})", flags: "g", confidence: 0.80 },
  ],

  health_condition: [
    // Diagnosis / condition labels
    { pattern: "(?:Diagnosis|Condition|Disorder|Disease)[:\\s]+[A-Za-z][A-Za-z\\s\\-]{3,50}", flags: "gi", confidence: 0.78 },
    { pattern: "(?:diagnosed\\s+with|suffers\\s+from|history\\s+of)[:\\s]+[A-Za-z][A-Za-z\\s\\-]{3,50}", flags: "gi", confidence: 0.82 },
  ],

  race: [
    {
      pattern:
        "\\b(?:African[- ]American|Black|White|Caucasian|Asian|Hispanic|Latino|Latina|Pacific Islander|Native American|Indigenous|Multiracial|Mixed[- ]race)\\b",
      flags: "gi",
      confidence: 0.80,
    },
    { pattern: "(?:Race|Ethnicity)[:\\s]+[A-Za-z][A-Za-z\\s\\-]{2,30}", flags: "gi", confidence: 0.85 },
  ],

  religion: [
    {
      pattern:
        "\\b(?:Christian|Catholic|Protestant|Muslim|Jewish|Hindu|Buddhist|Sikh|Atheist|Agnostic|Mormon|LDS|Evangelical)\\b",
      flags: "gi",
      confidence: 0.78,
    },
    { pattern: "(?:Religion|Religious\\s+belief)[:\\s]+[A-Za-z][A-Za-z\\s\\-]{2,30}", flags: "gi", confidence: 0.88 },
  ],

  political_opinion: [
    { pattern: "(?:Political\\s*(?:affiliation|party|opinion|view))[:\\s]+[A-Za-z][A-Za-z\\s\\-]{2,30}", flags: "gi", confidence: 0.87 },
  ],

  sexual_orientation: [
    {
      pattern:
        "\\b(?:gay|lesbian|bisexual|transgender|non-binary|queer|heterosexual|straight|asexual|LGBTQ\\+?)\\b",
      flags: "gi",
      confidence: 0.82,
    },
    { pattern: "(?:Sexual\\s*orientation)[:\\s]+[A-Za-z][A-Za-z\\s\\-]{2,30}", flags: "gi", confidence: 0.90 },
  ],

  biometric: [
    { pattern: "(?:fingerprint|retinal scan|facial recognition|voice print|biometric)[:\\s]+[A-Za-z0-9]{4,64}", flags: "gi", confidence: 0.85 },
  ],

  genetic: [
    { pattern: "(?:genetic|DNA|genomic)[:\\s]+(?:data|sequence|result|variant)[:\\s]+[A-Za-z0-9]{4,64}", flags: "gi", confidence: 0.85 },
    { pattern: "\\b(?:BRCA1|BRCA2|HER2|TP53|APOE)\\b", flags: "gi", confidence: 0.88 },
  ],

  union_membership: [
    { pattern: "(?:Union|Labor\\s*union|Trade\\s*union)\\s*(?:member|membership)[:\\s]+[A-Za-z][A-Za-z\\s\\-]{2,40}", flags: "gi", confidence: 0.85 },
  ],

  device_id: [
    // MAC address
    { pattern: "\\b([0-9A-Fa-f]{2}[:\\-]){5}[0-9A-Fa-f]{2}\\b", flags: "g", confidence: 0.95 },
    // UUID/GUID
    { pattern: "\\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\\b", flags: "g", confidence: 0.90 },
  ],

  voter_id: [
    { pattern: "(?:Voter\\s*ID|Voter\\s*Registration)[:\\s#]*[A-Z0-9]{6,20}", flags: "gi", confidence: 0.87 },
  ],
};

// ---------------------------------------------------------------------------
// Detection function
// ---------------------------------------------------------------------------

/**
 * Detects PII in the given text for the specified categories.
 * Returns sorted, non-overlapping matches above the confidence threshold.
 */
export function detectPii(
  text: string,
  target: PiiTarget,
): PiiMatch[] {
  const confidenceThreshold = target.confidence_threshold ?? 0.80;
  const categories = expandCategories(target.categories);
  const allMatches: PiiMatch[] = [];

  for (const category of categories) {
    const entries = PATTERNS[category as keyof typeof PATTERNS];
    if (!entries) continue;

    for (const entry of entries) {
      if (entry.confidence < confidenceThreshold) continue;
      const re = new RegExp(entry.pattern, entry.flags);
      let match: RegExpExecArray | null;
      while ((match = re.exec(text)) !== null) {
        // Prevent infinite loops on zero-length matches
        if (match[0].length === 0) {
          re.lastIndex++;
          continue;
        }
        allMatches.push({
          category,
          start: match.index,
          end: match.index + match[0].length,
          original: match[0],
          confidence: entry.confidence,
        });
      }
    }
  }

  // Sort by start position, then resolve overlaps (keep highest confidence)
  return resolveOverlaps(allMatches);
}

/**
 * Redacts detected PII spans in text, replacing each with [REDACTED:<category>].
 * Returns the redacted text and the list of applied matches.
 */
export function redactText(text: string, matches: PiiMatch[]): string {
  if (matches.length === 0) return text;
  let result = "";
  let cursor = 0;
  for (const match of matches) {
    result += text.slice(cursor, match.start);
    result += `[REDACTED:${match.category.toUpperCase()}]`;
    cursor = match.end;
  }
  result += text.slice(cursor);
  return result;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/** Removes overlapping matches, keeping the highest-confidence one per span. */
function resolveOverlaps(matches: PiiMatch[]): PiiMatch[] {
  if (matches.length === 0) return [];

  // Sort by start, then by confidence desc
  const sorted = matches.slice().sort((a, b) => {
    if (a.start !== b.start) return a.start - b.start;
    return b.confidence - a.confidence;
  });

  const result: PiiMatch[] = [];
  let lastEnd = -1;

  for (const match of sorted) {
    if (match.start >= lastEnd) {
      result.push(match);
      lastEnd = match.end;
    } else if (match.confidence > (result[result.length - 1]?.confidence ?? 0)) {
      // Higher confidence match that overlaps — replace the last one
      result.pop();
      result.push(match);
      lastEnd = match.end;
    }
  }

  return result;
}
